crypto/multisig/src/v2.rs

```diff
@@ -0,0 +1,1123 @@
+)]
+pub struct SignerContainer<S: Signer> {
+    /// The underlying signer entity.
+    /// This is made optional because of how Prost works. It will be None for
+    /// unsupported tags.
+    #[prost(oneof = "SignerEntity", tags = "1, 2")]
+    pub entity: Option<SignerEntity<S>>,
+}
+
+impl<S: Signer> From<SignerEntity<S>> for SignerContainer<S> {
+    fn from(entity: SignerEntity<S>) -> Self {
+        Self {
+            entity: Some(entity),
+        }
+    }
+}
+
+impl<S: Signer> From<S> for SignerContainer<S> {
+    fn from(signer: S) -> Self {
+        Self {
+            entity: Some(SignerEntity::Single(signer)),
+        }
+    }
+}
+
+impl<S: Signer> From<SignerSetV2<S>> for SignerContainer<S> {
+    fn from(signer_set: SignerSetV2<S>) -> Self {
+        Self {
+            entity: Some(SignerEntity::Multi(signer_set)),
+        }
+    }
+}
+
+/// A set of M-out-of-N signers (either individual signers or m-out-of-n groups
+/// of signers).
+#[derive(
+    Clone, Deserialize, Digestible, Eq, Hash, Message, Ord, PartialEq, PartialOrd, Serialize,
+)]
+pub struct SignerSetV2<S: Signer> {
+    #[prost(message, repeated, tag = "1")]
+    signers: Vec<SignerContainer<S>>,
+
+    #[prost(uint32, tag = "2")]
+    threshold: u32,
+}
+
+impl<S: Signer> SignerSetV2<S> {
+    /// Construct a new `SignerSetV2` from a list of signers and threshold.
+    pub fn new(signers: Vec<SignerContainer<S>>, threshold: u32) -> Self {
+        Self { signers, threshold }
+    }
+
+    /// Get the list of potential signers.
+    pub fn signers(&self) -> &[SignerContainer<S>] {
+        &self.signers
+    }
+
+    /// Get the threshold.
+    pub fn threshold(&self) -> u32 {
+        self.threshold
+    }
+
+    /// Verify a message against a multi-signature, returning the list of
+    /// signers that signed it.
+    ///
+    /// NOTE: If a signer appears in multiple groups, it will only be matched
+    /// once, and for the first group it appears in that passed the signing
+    /// threshold (and the ordering of the groups and signatures depends on
+    /// their byte representation since we always sort them). Here is an
+    /// example that demonstrates the implications of this: Assume a signer set,
+    /// that requires 1-of-2 and is sorted, and where each member is a
+    /// m-of-n group where the first requires 2/3 and the second requires
+    /// 3/3:
+    /// 1) A, B, C (threshold 2)
+    /// 2) A, D, E (threshold 3)
+    /// If we are passed signatures A, B, C, D, E we could in theory satisfy the
+    /// threshold of both groups but since A gets "consumed" when matching the first group,
+    /// leaving the second group with only D and E, it will not be
+    /// considered a match. We say we could match all 5 signatures
+    /// in theory, because if we ignored A for the first set, it could still be
+    /// used to match the 2nd set. Such algorithm would require a more
+    /// complex (and slower) implementation, and is not currently supported.
```

Comment:
I think this is similar to https://en.wikipedia.org/wiki/Set_cover_problem?

Reply:
I think it's also similar to our quorum set implementations, right?
Our quorum set stuff allows that different quorum sets can overlap and it's okay.

Reply:
It is slightly different since in QuorumSets you are okay with a single node id being matched in multiple nested definitions, but when dealing with signatures, following a discussion that took place when the original `mc-crypto-multisig` crate was built, we want to not allow signers/signatures to be matched more than once.

That is to protect against attacks where someone finds a way to generate multiple signatures that look differently but are all by the same signer, or figuring out a way to have a single signature match multiple signers. Both are very unlikely, but it was mentioned that it's better to just not allow that.

Reply:
> generate multiple signatures that look differently but are all by the same signer

I understand that this is a "general issue" with ed25519 but it doesn't seem relevant to this code, since we are counting which public keys the signature matched against, and not actually counting a number of signatures.

> figuring out a way to have a single signature match multiple signers

I don't believe that this is possible unless the signing keys are chosen in a pathological way, so I don't think this is a significant consideration. I would be happy to revisit the initial discussion, because maybe something was lost in communication.

Reply:
I think the implementation we have is going to create a lot of difficulties in practice because it will be hard to figure out why a set of signatures doesn't satisfy the multisig. I think it's possible that we are chasing our tails and making things much more complicated than they need to be.

Reply:
Having overlapping signing sets is fine and allowed, it only gets weird when you have stuff like in the comment where a signer got "counted" towards a SignerSet's threshold where it wasn't strictly necessary since that threshold was exceeded by other signers, and we need this singer in a different SignerSet.

I'm not opposed to allowing the same signer to match multiple times, but it means the following configuration might result in undesirable behavior - depending on our own definition of what is desirable:
```
S1 = [1 out of A, B]
S2 = [1 out of A, C]
S3 = [2 out of S1, S2]
```
This is a questionable configuration that can be interpreted in two ways:
1) `A` satisfies both S1 and S2 so providing only A is sufficient 
2) We want either A+C or A+B
The right way for that, is to structure it as 2 out of 2 of `[A, [B, C]]` - so we could just say we don't care about configuration that looks like the 2nd one.

Reply:
When I see a thing like `S3 = [2 out of S1, S2]`, what I expect that to mean intuitively is?

First, was S1 satisfied by the given signature set for given message?
Next, was S2 satisfied by the given signature set for given message?

If so, then S3 is satisfied, otherwise it isn't.

And I expect that these first two questions make sense and can be answered in isolation of one another.
If the actual implementation is stateful and tries to prevent A from being used in both sets, then "S3 = [2 out of S1, S2]" doesn't actually mean what I as a user expect, because it becomes possible that it is true that "S2 is satisfied by a given set of signatures for a given message", and calling `.verify` with `S2`, that set of signatures, and that message, returns true, but that when we evaluate it in the context of S3, we decide that it isn't satisfied.

I think we should consider such discrepancy to be a bug. So that's why I think we should get rid of any extra state that influences this check, in the form of the "seen_signers" parameter.

Reply:
Works for me